merino/featureflags.py

```
@@ -0,0 +1,107 @@
+    of the request as the bucketing key so that feature checks within a given search session would
+    be consistent. Additionally you can pass a custom bucketing_id via the `bucket_for` parameter.
+    This is useful when you have an ad-hoc bucketing identifier that is not supported via one of
+    the standard schemes.
+
+    Each flag has a very simple configuration:
+
+    ```
+    [default.flags.<flag_name>]
+    scheme = 'session'
+    enabled = 0.5
+    ```
+
+    `scheme` - This is the bucketing scheme for the flag. Allowed values are 'random' and 'session'
+    `enabled` - This represents the % enabled for the flag and must be a float between 0 and 1
+    """
+
+    flags: dict
+
+    def __init__(self) -> None:
+        self.flags = _flags.get("flags", {})
+
+    def is_enabled(self, flag: str, bucket_for: str | bytes | None = None) -> bool:
+        """
+        Checks if a given flag is enabled via a feature flag configuration block.
+        As a principal it fails closed as quickly as possible.
+        """
+        config = self.flags.get(flag)
+        if config is None:
+            return False
+
+        enabled = config.get("enabled", 0.0)
+        # enabled should be gt 0 and lt 1
+        if enabled <= 0.0 or enabled > 1.0:
+            return False
+
+        bucket_scheme = config.get("scheme")
+        if (bucketing_id := self._get_bucketing_id(bucket_scheme, bucket_for)) is None:
+            return False
+
+        bucket_value = self._bytes_to_interval(bucketing_id)
+        return bucket_value <= enabled
+
+    def _get_bucketing_id(
+        self, scheme: str, bucket_for: str | bytes | None
+    ) -> bytes | None:
+        # override bucketing id if specified in args
+        if bucket_for is not None:
+            bucket_type = type(bucket_for)
+            if bucket_type is str:
+                return self._get_digest(cast(str, bucket_for))
+            elif bucket_type is bytes:
+                return cast(bytes, bucket_for)
```

Shall we throw over here when seeing an invalid `bucket_for` type rather than return `None` implicitly? If we expect it to be either str or bytes if present, having other types sounds like a bug to me, and I think we should fail loud.

---

We could throw here but I wouldn't want those exceptions to bubble up past `is_enabled`. I don't want `is_enabled` checks to ever throw. They should fail to False to prevent extra branching around catching exceptions for every flag check. I would be down with adding logging or metrics in situations like this. that would allow us to have visibility into issues